kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/SchemaPrinter.kt

```diff
@@ -0,0 +1,290 @@
+                if (index > 0) {
+                    appendLine()
+                }
+                appendDescription(type)
+                val possibleTypes = type.possibleTypes.sortedByName().map { it.name }
+                appendLine("union ${type.name} = ${possibleTypes.joinToString(" | ")}")
+            }
+        }
+
+        // Enums
+        //  https://spec.graphql.org/draft/#sec-Enums
+        val enums = buildString {
+            schemaTypes.filter { it.kind == TypeKind.ENUM }.forEachIndexed { index, type ->
+                if (index > 0) {
+                    appendLine()
+                }
+                appendDescription(type)
+                appendLine("enum ${type.name} {")
+                val indentation = "  "
+                type.enumValues.sortedByName().forEach { enumValue ->
+                    appendDescription(enumValue, indentation)
+                    appendLine("${indentation}${enumValue.name}${enumValue.deprecationInfo()}")
+                }
+                appendLine("}")
+            }
+        }
+
+        // Input Types
+        //  https://spec.graphql.org/draft/#sec-Input-Objects
+        val inputTypes = buildString {
+            schemaTypes.filter { it.kind == TypeKind.INPUT_OBJECT && !it.inputFields.isNullOrEmpty() }
+                .forEachIndexed { index, type ->
+                    if (index > 0) {
+                        appendLine()
+                    }
+                    appendDescription(type)
+                    appendLine("input ${type.name}${type.implements()} {")
+                    val indentation = "  "
+                    type.inputFields.sortedByName().forEach { field ->
+                        appendDescription(field, indentation)
+                        appendLine("${indentation}${field.name}: ${field.type.typeReference()}${field.deprecationInfo()}")
+                    }
+                    appendLine("}")
+                }
+        }
+
+        // Directives
+        //   https://spec.graphql.org/draft/#sec-Type-System.Directives.Built-in-Directives
+        //   "When representing a GraphQL schema using the type system definition language any built-in directive may be omitted for brevity."
+        val directives = buildString {
+            schema.directives.filter { config.includeBuiltInDirectives || !it.isBuiltIn() }.sortedByName()
+                .forEachIndexed { index, directive ->
+                    if (index > 0) {
+                        appendLine()
+                    }
+                    val args = directive.args.takeIf { it.isNotEmpty() }
+                        ?.joinToString(", ", prefix = "(", postfix = ")") { arg ->
+                            arg.name + ": " + arg.type.typeReference() + arg.defaultInfo()
+                        } ?: ""
+                    appendLine("directive @${directive.name}$args on ${directive.locations.joinToString(" | ") { it.name }}")
+                }
+        }
+
+        return listOf(
```

Comment:
We could also let all optional stuff here be null instead of an empty string and use `listOfNotNull` here. IMO `null` is a better representation of the absence of something than a blank string.

Reply:
Agreed. I think there must have been a reason for implementing it this way, let me double-check and fix if possible. Might have been from some intermediate version.

Reply:
Ah. Most of the stuff is wrapped in `buildString` so it defaults to an empty string. Not sure if it would be a real improvement to append `.takeIf { it.isNotBlank() }` everywhere.